### DISCUSSION TITLE
README notes

### DISCUSSION THREAD

**Discussion by harens** [beartype/beartype#28] [General]
First discussion yay! 🥳 

I was hoping that this discussion could be used to note small corrections and to add useful resources to help us improve the README.

I've noticed that quite a few commits recently have been related to the README. Until the time comes when it gets broken down into chunks as part of RTD, I thought I'd start a discussion as to how we can make it even better (baby steps at a time people 👶).

As a starting point, I noticed a few things that don't really warrant a PR. They can just be added in with another commit.

* The line `def __init__(self, scalar: ScalarTypes) -> NoneType:` generates a mypy error: ` error: The return type of "__init__" must be None`.
    * Following https://github.com/beartype/beartype/commit/ffc233f1df0525aa881820f103098a7f9366f10f, I presume this can just be replaced by `None`.
* The code blocks added by https://github.com/beartype/beartype/commit/00ec651f70ae159e07876269c9226b30fb299242 have been indented. This gives it the "quote" effect, and it doesn't render the code properly on the GitHub app.
    * There are also [some other code blocks](https://github.com/beartype/beartype#industrial-example) that have this issue. 

_(GitHub does not make it easy to reference specific lines in rst files, so apologies about that)._
 
Even though your writing is truly amazing ✨, I think the README is slightly too long....and we might find [this article](https://skerritt.blog/make-popular-open-source-projects/) by the amazing @bee-san useful in the future (I certainly did). But, as I said, baby steps at a time and we'll make good progress.

**Comment by leycec** [comment]
**Ugh!** You continue to amaze even as our documentation continues to appal. I've been relentlessly putting out [blazing dumpster fires across the codebase with my own burning body](https://github.com/beartype/beartype/issues/29#issuecomment-789508939), culminating in this tragic reenactment of [*Home Trepanation for Dummies*](https://en.wikipedia.org/wiki/Trepanning):

![My head after solving #19.](https://camo.githubusercontent.com/45f18e4640b2a85c953b9868eb41f02bbe9b1f2f9bed5bb94b01c9140f10a63f/68747470733a2f2f6d65646961696e7265766965772e66696c65732e776f726470726573732e636f6d2f323031362f30332f686561642d6578706c6f64696e672e706e673f773d36303026683d343530)
<sup>*@leycec: he's looked better*</sup>

I narrowly avoided rupturing my brain.

Now that issue #19 has been conquered with bold justice and righteous wrath, we stand on the blood-dimmed tidelines of a new `beartype` release. It's corpulent with features, so we've upgraded it from a mere `0.5.2` point release to a major `0.6.0` minor release. I'm pushing it out tonight before I lose any further sanity points, **complete with the horrifyingly accurate corrections you've noted above.**

From your elevated stature as a manly genius, @harens, `beartype 0.6.0` will both excite and thrill you with its:

* **``None`` support.** *Yeah.* That shouldn't have taken us until `beartype 0.6.0`, but here we are. Coders gonna code.
* **PEP 561 compliance,** which we've miraculously preserved for at least one week due to integrating mypy into our test suite, which now proudly fails whenever we violate mypy. That's several times a day.

> Even though your writing is truly amazing :sparkles:, I think the `README` is slightly too long....

Your expertise as a build automation engineer is matched only by your propensity for gratuitous ~~lies~~ kindness, for which I am thankful.

Let's be clear: **the `README` is balls.** We all know it's balls, and that's okay. I'm only mildly ashamed of what I have done, and that's probably not okay.

My first real-world goals after releasing `beartype 0.6.0` are:

1. Take a six-hour shower in a bold reimagining of [80's-era Irish Spring commercials](https://www.youtube.com/watch?v=VjSNrg7T0Wo), complete with exaggerated and frankly insulting Dublin accent.
1. Publish a GitHub Sponsors profile, complete with offensive begging and a staged photograph of @leycec in tattered beggar's rags sprawled helplessly in a snowbank.
1. **Split our GitHub-hosted `README.rst` file into multiple ReadTheDocs-hosted reST files in our `doc/` subdirectory,** because the current documentation em**bear**asses us all.

<sup>I'll show myself out now.</sup>

**Reply by harens** [reply]
> None support. Yeah. That shouldn't have taken us until beartype 0.6.0, but here we are. Coders gonna code.

> PEP 561 compliance, which we've miraculously preserved for at least one week due to integrating mypy into our test suite, which now proudly fails whenever we violate mypy. That's several times a day.

The two things I couldn't have asked for more. It's an early Christmas 🎄. Thank you @leycec (and @Heliotrop3 😃) for your work on this project!

> Let's be clear: the README is balls. We all know it's balls, and that's okay. I'm only mildly ashamed of what I have done, and that's probably not okay.

Do you know how I found beartype? I was browsing [GitHub Explore](https://github.com/explore) and I saw this really interesting type checking project. After clicking on the link, I was amazed by the quality of writing, the witty jokes mixed in with some hilarious bear puns 🐻, and the dedication to write such thorough and detailed documentation. 

As I mention in my first beartype issue https://github.com/beartype/beartype/issues/12:

> I was so impressed by @leycec's way of writing issues/README/etc. (it's amazing ✨) that I felt compelled to help in the way I knew best, macOS package management.

One month later, and here we are. So...although we can certainly work on the README, always remember that it's astoundingly stupefying. 😀

**Comment by leycec** [comment]
Trivial `code-block` misrendering issues **resolved** by 3dd551e. (It is good.)

Non-trivial `README.rst` disembowelling  deferred to a day when I have slept.

Thanks for all the documentation assurance, @harens. I remain assured you will someday be the Prime Minister of Scotland and that, as a true-blooded Scotsman, I will have to do what you say. Luckily, that day is not today. :scotland:

**Comment by leycec** [comment]
> Do you know how I found beartype? I was browsing GitHub Explore and I saw this really interesting type checking project. After clicking on the link, I was amazed by the quality of writing, the witty jokes mixed in with some hilarious bear puns :bear:, and the dedication to write such thorough and detailed documentation.

Awwwww. @harens and @leycec BGHCFRLs (Best GitHub Coders for Repository Life)!

But seriously. This is a serious true-life story that I should publicly market in a new **Stories** section of our documentation – complete with big Unicode quotes, flashy HTML5 animations, and a plucky selfie of @harens injecting @beartype all over [his seaport](https://github.com/harens/seaport).

On second thought... *maybe we have enough documentation already.*

**Comment by langfield** [comment]
Perhaps this is a controversial take, but I have to say having all the documentation on one searchable page is one of the things I like most about beartype vs. other projects. When I need to know something, I go to the main page of the github repository and just `Ctrl+F`. And that's it. Incredibly simple. Love it.

ReadTheDocs sites tend to suffer from an orphan problem, where there'll be links in the sidebar about something you're looking for, and you go there to find the page hasn't been written yet.

That being said, having some sort of autogenerated API docs would be sweet.

**Comment by leycec** [comment]
> 'Where are the new frontiers?' the Romantics cried, unaware that the frontier of the mind had opened.

> — Alfred Bester, *The Stars My Destination*

Here I quote the greatest sci-fi novel nobody ever read by the greatest sci-fi author nobody ever read. Although this has no relation whatsoever to anything written above, I feel it impingent upon my soul as a sentient ape to highlight this good thing written by another sentient ape. Please do:

* Read *The Stars My Destination.*
* Not read anything about *The Stars My Destination* before doing so.

This transmission has been sent to you by unseen forces of universal goodwill. :space_invader: :man_astronaut: :artificial_satellite: 

> ...having all the documentation on one searchable page is one of the things I like most about beartype vs. other projects.

`\o/`

**Yes!** This is the Craigslist approach to user experience (UX), where the usability of web design is directly proportionality to its ugliness and your desire to kick the responsible designer in the shins repeatedly with a burlap sack filled with fermented compost.

> When I need to know something, I go to the main page of the github repository and just `Ctrl+F`. And that's it. Incredibly simple. Love it.

Discoverability. Laziness. These two things intersect in @beartype.

When we do inevitably facelift our docos with RtD-hosted botox, preserving cheap and easy discoverability is our foremost priority. There's [actually a quiet discussion about this from the Sphinx perspective here](https://github.com/readthedocs/readthedocs-sphinx-search/issues/87). The third-party `readthedocs-sphinx-search` Sphinx extension (which we don't use, but we *could*) binds the `/` key ala Vim to ignite a search across RtD-hosted docos. That seems like a responsible compromise.

We promise you this: @beartype docos will *never* hijack `<Ctrl-f>` for its own malignant purposes or my username isn't `bonoboheartbanana`.

> ReadTheDocs sites tend to suffer from an orphan problem...

**Yeah.** Let's make sure @leycec never does that. Burlap sacks filled with fermented compost are a credible threat here.

> ...having some sort of autogenerated API docs would be sweet.

![sweat face](https://media.tenor.com/NE5HH5aAmPcAAAAS/hop-pop-plantar-amphibia.gif)

Shockingly, we actually *did* get `autoapi` wired up this summer — but then things got cray-cray in both @beartype land and in real life. We never finished the job. It's possible that API documentation is currently being correctly generated and just not linked to anywhere. I... should really look into that.

This is an awkward moment for us all.

**Reply by posita** [reply]
> > 'Where are the new frontiers?' the Romantics cried, unaware that the frontier of the mind had opened.
> 
> > — Alfred Bester, _The Stars My Destination_
> 
> Here I quote the greatest sci-fi novel nobody ever read by the greatest sci-fi author nobody ever read. Although this has no relation whatsoever to anything written above, I feel it impingent upon my soul as a sentient ape to highlight this good thing written by another sentient ape. ...
> 
> This transmission has been sent to you by unseen forces of universal goodwill. :space_invader: :man_astronaut: :artificial_satellite:

My horizons continue to be broadened by your literature recommendations. I just finished my _first_ reading of _The Stars My Destination_ late last night (way too late, but I had to finish, may my spouse forgive me). Another choice quote (also not relating to any of the above, at least not in ways any of us would immediately acknowledge):

> Commerce obeyed, for this war (like all wars) was the shooting phase of a commercial struggle.

Priceless.
